lib/rosette/integrations/smartling-integration/smartling_completer.rb

```diff
@@ -1,49 +1,164 @@
-            completed_files.each do |file|
-              file_uris_to_files[file.file_uri] ||= file
-              file_hash[file.file_uri] << locale
+        def complete
+          build_completion_maps.each do |file_uri, locales|
+            if locales.all? { |_, v| v.complete? }
+              begin
+                file = locales[repo_config.locales.first.code]
+                complete_log(file)
+                delete_file(file)
+              rescue => e
+                rosette_config.error_reporter.report_error(e)
+              end
             end
           end
+        end
 
-          file_hash.each do |file_uri, completed_locales|
-            if completed_locales.sort == locales.sort
-              Rosette.logger.info("Deleting file from Smartling: #{file_uri}")
-              file = file_uris_to_files[file_uri]
+        private
 
-              configuration.datastore.add_or_update_commit_log(
-                file.repo_name, file.commit_id, nil, Rosette::DataStores::PhraseStatus::TRANSLATED
-              )
+        def delete_file(file)
+          Retrier.retry(times: 3) do
+            smartling_api.delete(file.file_uri)
+          end.on_error(Exception).execute
+        end
 
-              smartling_apis[repo_name].delete(file_uri)
+        def complete_log(file)
+          rosette_config.datastore.add_or_update_commit_log(
+            file.repo_name, file.commit_id, nil, Rosette::DataStores::PhraseStatus::TRANSLATED
+          )
+        end
+
+        def build_completion_maps
+          if thread_pool_size > 1
+            build_completion_maps_asynchronously
+          else
+            build_completion_maps_synchronously
+          end
+        end
+
+        def build_completion_maps_synchronously
+          {}.tap do |completion_map|
+            repo_config.locales.each do |locale|
+              cur_map = build_completion_map_for(locale)
+              merge_completion_maps!(completion_map, cur_map)
             end
           end
         end
 
-        private
+        def build_completion_maps_asynchronously
+          completion_map = {}
+          pool = Concurrent::FixedThreadPool.new(thread_pool_size)
+          merge_mutex = Mutex.new
+
+          repo_config.locales.each do |locale|
+            pool << Proc.new do
+              cur_map = build_completion_map_for(locale)
+
+              merge_mutex.synchronize do
+                merge_completion_maps!(completion_map, cur_map)
+              end
+            end
+          end
+
+          drain_pool(pool, repo_config.locales.size)
+          completion_map
+        end
+
+        def drain_pool(pool, total)
+          pool.shutdown
+          last_completed_count = 0
+
+          while pool.shuttingdown?
+            current_completed_count = pool.completed_task_count
+
+            if current_completed_count > last_completed_count
+              puts "#{current_completed_count} of #{total} completed"
```

Comment:
We do want to use the logger, good catch :)

Keeping track of the merged maps is a cool idea, I'll see if I can refactor things a bit.